docs/proposals/repo-server-multi-tenancy.md

```diff
@@ -0,0 +1,200 @@
+
+The relationship of an `AppProject` to a repository server instance is `1:n`. An `AppProject` instance can use exactly one repository server instance, while each repository server instance can be referenced by multiple `AppProjects`. 
+
+### Changes to caching
+
+Redis is used as a central instance to cache manifests generated by the repository server. The manifest cache currently uses keys that do not consider multiple independent repository servers storing manifests in the same cache. Also, by default there is no authentication between the repository server and the cache, and access control is solely implemented using network policies.
+
+There are a couple of challenges to be solved with the cache:
+
+1. Repository server instances must be able to make distinct cache entries, possibly for the same repositories and same revisions.
+1. Any repository server instance must not be allowed to gain access to cache entries from another repository server instance
+
+In initial discussions, a possible solution for the first challenge would be to prefix cache keys with the instance name of the repository server (e.g. `instance-1`). This is easy to implement, and only requires minimal changes and efforts to the code.
+
+For the second challenge however, there have been multiple ideas of various complexity.
+
+**1. Using authentication and ACLs in Redis**: This would involve setting up users and ACLs in Redis, and augmenting repository server instance configuration with credentials. This is however, complex to setup and maintain and could also prevent some users from using external (managed) Redis instance where this is not supported or would cost additional money.
+
+**2. Moving the manifest cache to the application controller:** Another idea is to move the handling of the cache completely to the application controller, and remove the relationship between repository server and the cache. This would also save some network traffic between the application controller and the repository server, which would be especially efficient when the application controller requests manifests from the repository server and the repository server would just return the cached manifests.
+
+**3. Encryption of cache entries**: The third idea revolves around encrypting the actual manifests stored in the cache, using a static encryption key unique to each repository server instance. This would prevent other repository server instances to gain manifests from other instances, ensuring confidentiality of the manifests in the cache (which would especially benefit manifests that are confidential and are decrypted on generation time, such as git-crypt). This would come at cost of some compute resources, but probably be worth it.
```

Comment:
I think we could probably use some details about the generation and storage of the key.

Reply:
Good thinking.

My initial idea was to handle generation and storage similar to how we handle `server.secretKey` right now, i.e. generate it at startup (in the application controller) and store it in the `argocd-secret`. 

Then, to have the repository server use the key, we could supply it in the gRPC calls we make from API server and application controller. This would have the advantage that the repository server doesn't need to store or persist the key anywhere (which would make the key susceptible to being stolen), but could use it once it needs to access the cache (e.g. in case if instructed by another component).

This approach would also enable future use-cases, such as per AppProject encryption keys and keep propagation of key in case of rotation rather simple.

We could design something like this, if there's not a serious hole in my logic that I've not seen.